### DISCUSSION TITLE
Error in reading new files -pyPrimemesh

### DISCUSSION THREAD

**Discussion by som1197** [ansys/pyprimemesh#817] [Q&A]
Hello,

I am using pyPrimemesh as a client with my ANSYS 23 R2 version located in a different drive (D) instead of the default. I am unable to read any files (Default examples and my .scdoc designs). This is the error that I get after reading one of my files.

```
import os
import tempfile

from ansys.meshing import prime
from ansys.meshing.prime.graphics import Graphics

prime_client = prime.Client(ip = "10.0.0.41", port=50055, timeout = 60)
model = prime_client.model

mesh_util = prime.lucid.Mesh(model=model)
file_name = r'..\case_3a.1.scdoc'
params = prime.ImportCadParams(
    model=model, cad_reader_route=prime.CadReaderRoute.SPACECLAIM
)
prime.FileIO(model).import_cad(file_name=file_name, params=params)
```

```
---------------------------------------------------------------------------
PrimeRuntimeError                         Traceback (most recent call last)
Cell In[16], line 4
      1 params = prime.ImportCadParams(
      2     model=model, cad_reader_route=prime.CadReaderRoute.SPACECLAIM
      3 )
----> 4 prime.FileIO(model).import_cad(file_name=file_name, params=params)

File ~\AppData\Local\anaconda3\envs\Ansys_related\Lib\site-packages\ansys\meshing\prime\core\fileio.py:504, in FileIO.import_cad(self, file_name, params)
    461 r"""Import a CAD file from disk.
    462 
    463 Supported CAD file formats on Windows are:
   (...)
    501 
    502 """
    503 with utils.file_read_context(self._model, file_name) as temp_file_name:
--> 504     import_result = super().import_cad(temp_file_name, params)
    505     if import_result.error_code == ErrorCode.NOERROR:
    506         self._model._sync_up_model()

File ~\AppData\Local\anaconda3\envs\Ansys_related\Lib\site-packages\ansys\meshing\prime\autogen\fileio.py:650, in FileIO.import_cad(self, file_name, import_cad_params)
    648 command_name = "PrimeMesh::FileIO/ImportCAD"
    649 self._model._print_logs_before_command("import_cad", args)
--> 650 result = self._comm.serve(self._model, command_name, self._object_id, args=args)
    651 self._model._print_logs_after_command("import_cad", ImportCadResults(model = self._model, json_data = result))
    652 return ImportCadResults(model = self._model, json_data = result)

File ~\AppData\Local\anaconda3\envs\Ansys_related\Lib\site-packages\ansys\meshing\prime\internals\error_handling.py:535, in error_code_handler.<locals>.decorator_error_code.<locals>.wrapper_error_code(*args, **kwargs)
    529     if error_code > 0:
    530         error_location_msg = (
    531             f'\nError Locations: {error_locations}'
    532             if len(error_locations) > 0
    533             else f''
    534         )
--> 535         raise PrimeRuntimeError(
    536             prime_error_messages.get(
    537                 ErrorCode(error_code), f'Unrecogonized error code {error_code}'
    538             )
    539             + error_location_msg,
    540             ErrorCode(error_code),
    541             error_locations,
    542         )
    544 prime_warnings = []
    545 single_warning = result.get('warningCode', None)

PrimeRuntimeError: Incorrect File Path or Name. Check your file name and path.
```

**Comment by waltersma** [comment]
Hi @som1197 ,

Glad to hear your first issue is resolved.  Please let us know what you did as others might face the same issue.

Regarding this issue the error states that the file cannot be found.  Have you confirmed the path is valid?

For instance:

```print(file_name, "is valid:\n", os.path.exists(file_name))```

**Reply by som1197** [reply]
Hi @waltersma ,

Yes, I will record the steps required to resolve the previous issue. Regarding this issue, I confirmed that my file path is valid.

Even if I use the file name of the default 'mixing elbow' example, it returns the same error.

To give some more context, my client side folder looks something like this.

![image](https://github.com/ansys/pyprimemesh/assets/46930662/a24164f0-5447-4c3b-8a67-c254834041d9)

On the server side, it looks something like this.

![image](https://github.com/ansys/pyprimemesh/assets/46930662/f6899828-5bb7-4fff-9b00-280a69a78de6)

When I used the code snippet for the mixing elbow case from the documentation, it downloaded those files at this location, where my anaconda installation is based. 

![image](https://github.com/ansys/pyprimemesh/assets/46930662/d1d48525-f393-4040-b018-bd10d6d223de)

I don't know if this information helps in any way, but this is something which I am specifically encountering on the client side. One the server side, everything works smoothly.

**Comment by rchopade7** [comment]
@som1197 can you try absolute file path instead of relative file path
![image](https://github.com/ansys/pyprimemesh/assets/163139281/25f7affb-9e35-4661-b8e3-524951a72955)

**Reply by som1197** [reply]
@rchopade7 , I tried that but it didn't work.

**Reply by som1197** [reply]
@waltersma or @rchopade7 , can you provide any pointers into resolving this?

**Reply by waltersma** [reply]
Hi @som1197, if the server is on a different machine can you place the files on a location that is accessible to the server?

Hope this helps!

**Reply by som1197** [reply]
Hi @waltersma, If I place the files at a location which I accessible to the server, I will have to use the Jupyter notebook on the server side. In my case, since I don't have any graphics hardware on the server, I am using my remote workstation. Even if I enter the file path of the server, since the code is on my local workstation, it wont parse any data.

**Reply by rchopade7** [reply]
@som1197 can you add below snippet before reading file on server
`import ansys.meshing.prime.internals.config as config`
`config.set_file_existence_check(False)`

**Reply by asherbranner** [reply]
> @som1197 can you add below snippet before reading file on server `import ansys.meshing.prime.internals.config as config` `config.set_file_existence_check(False)`

@rchopade7 @som1197 I apologize for my initial (now deleted) reply. 
This fix seems to work on my end. 
@som1197 can you verify that this works on your end when adding the provided code client-side?

**Reply by som1197** [reply]
Hello @rchopade7 , I used this snippet in my code and also placed the file being read in the current folder where my source code was located on the server side. It resolved my previous error, but cannot procure ANSYS SC prepost licenses to parse the file. I get the following error as shown in the image below when I try to parse the data using 4 processors.

![image](https://github.com/ansys/pyprimemesh/assets/46930662/a5b402c0-e078-4438-9f24-6be5f52b6a80)

I have 2 pre post licenses and it uses up one license while trying to parse the file but doesn't open up.

![image](https://github.com/ansys/pyprimemesh/assets/46930662/d307eff2-706d-4936-9a62-ac15d305ecbb)

**Reply by waltersma** [reply]
Hi @som1197 , please can you confirm my understanding:

You are running an Ansys 24R1 installation on the server machine and using 0.5.X PyPrimeMesh on the client machine.  The server machine does not have access to any licenses for SpaceClaim and you have 2 available cfd_preppost licenses before launching pyprimemesh and attempting to read a spaceclaim file.  Is this correct?

(Cc @prajaktmahajan )

**Reply by som1197** [reply]
Hi @waltersma 

Yes that is correct, and this is the error on the client side.
```
---------------------------------------------------------------------------
PrimeRuntimeError                         Traceback (most recent call last)
Cell In[3], line 1
----> 1 mesh_util.read('case_3a.1.scdoc',cad_reader_route=prime.CadReaderRoute.SPACECLAIM)

File ~\AppData\Local\anaconda3\envs\Ansys_software\Lib\site-packages\ansys\meshing\prime\lucid\mesh_util.py:92, in Mesh.read(self, file_name, append, cad_reader_route)
     88     prime.FileIO(self._model).read_pmdat(
     89         file_name, prime.FileReadParams(self._model, append=append)
     90     )
     91 else:
---> 92     prime.FileIO(self._model).import_cad(
     93         file_name,
     94         prime.ImportCadParams(
     95             self._model, append=append, cad_reader_route=cad_reader_route
     96         ),
     97     )

File ~\AppData\Local\anaconda3\envs\Ansys_software\Lib\site-packages\ansys\meshing\prime\core\fileio.py:504, in FileIO.import_cad(self, file_name, params)
    461 r"""Import a CAD file from disk.
    462 
    463 Supported CAD file formats on Windows are:
   (...)
    501 
    502 """
    503 with utils.file_read_context(self._model, file_name) as temp_file_name:
--> 504     import_result = super().import_cad(temp_file_name, params)
    505     if import_result.error_code == ErrorCode.NOERROR:
    506         self._model._sync_up_model()

File ~\AppData\Local\anaconda3\envs\Ansys_software\Lib\site-packages\ansys\meshing\prime\autogen\fileio.py:650, in FileIO.import_cad(self, file_name, import_cad_params)
    648 command_name = "PrimeMesh::FileIO/ImportCAD"
    649 self._model._print_logs_before_command("import_cad", args)
--> 650 result = self._comm.serve(self._model, command_name, self._object_id, args=args)
    651 self._model._print_logs_after_command("import_cad", ImportCadResults(model = self._model, json_data = result))
    652 return ImportCadResults(model = self._model, json_data = result)

File ~\AppData\Local\anaconda3\envs\Ansys_software\Lib\site-packages\ansys\meshing\prime\internals\error_handling.py:535, in error_code_handler.<locals>.decorator_error_code.<locals>.wrapper_error_code(*args, **kwargs)
    529     if error_code > 0:
    530         error_location_msg = (
    531             f'\nError Locations: {error_locations}'
    532             if len(error_locations) > 0
    533             else f''
    534         )
--> 535         raise PrimeRuntimeError(
    536             prime_error_messages.get(
    537                 ErrorCode(error_code), f'Unrecogonized error code {error_code}'
    538             )
    539             + error_location_msg,
    540             ErrorCode(error_code),
    541             error_locations,
    542         )
    544 prime_warnings = []
    545 single_warning = result.get('warningCode', None)

PrimeRuntimeError: CAD import failed.
```

**Comment by waltersma** [comment]
Hi @som1197, 

For CAD import where you need cad reader routes using other products you will need those products installed, configured and licensed on the machine you are using for the server.  In this example you are attempting to read a SpaceClaim SCDOC file on a machine without a license available to launch SpaceClaim.

In this case you can export to a native format that can be read without the need for this setup.  For example, if you export FMD format from SpaceClaim locally where you have a license you will be able to read this format into PyPrimeMesh on the server without needing a SpaceClaim license and product installed on the server using the native cad reader route.

Please let me know if this answers your question.

Thanks

(Cc @prajaktmahajan )

**Reply by som1197** [reply]
@waltersma @prajaktmahajan : So I exported the case_3a.1.scdoc as case_3a.1.fmd and stored it at the location where the SC license was available. Based on what I understood from your previous thread, it doesn't need to be parsed through the conventional SC route. But I still end up getting this error.

```
import os
import tempfile

from ansys.meshing import prime
from ansys.meshing.prime.graphics import Graphics

import ansys.meshing.prime.internals.config as config
config.set_file_existence_check(False)

prime_client = prime.Client(ip = "10.0.0.41", port=50055, timeout = 60)
model = prime_client.model
mesh_util = prime.lucid.Mesh(model=model)
mesh_util.read('case_3a.1.fmd')
```

```
---------------------------------------------------------------------------
PrimeRuntimeError                         Traceback (most recent call last)
Cell In[13], line 1
----> 1 mesh_util.read('case_3a.1.fmd')

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\ansys\meshing\prime\lucid\mesh_util.py:114, in Mesh.read(self, file_name, append, cad_reader_route)
    110     prime.FileIO(self._model).read_pmdat(
    111         file_name, prime.FileReadParams(self._model, append=append)
    112     )
    113 else:
--> 114     prime.FileIO(self._model).import_cad(
    115         file_name,
    116         prime.ImportCadParams(
    117             self._model, append=append, cad_reader_route=cad_reader_route
    118         ),
    119     )

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\ansys\meshing\prime\core\fileio.py:586, in FileIO.import_cad(self, file_name, params)
    543 r"""Import a CAD file from disk.
    544 
    545 Supported CAD file formats on Windows are:
   (...)
    583 
    584 """
    585 with utils.file_read_context(self._model, file_name) as temp_file_name:
--> 586     import_result = super().import_cad(temp_file_name, params)
    587     if import_result.error_code == ErrorCode.NOERROR:
    588         self._model._sync_up_model()

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\ansys\meshing\prime\autogen\fileio.py:692, in FileIO.import_cad(self, file_name, import_cad_params)
    690 command_name = "PrimeMesh::FileIO/ImportCAD"
    691 self._model._print_logs_before_command("import_cad", args)
--> 692 result = self._comm.serve(self._model, command_name, self._object_id, args=args)
    693 self._model._print_logs_after_command("import_cad", ImportCadResults(model = self._model, json_data = result))
    694 return ImportCadResults(model = self._model, json_data = result)

File ~\AppData\Local\Programs\Python\Python311\Lib\site-packages\ansys\meshing\prime\internals\error_handling.py:577, in error_code_handler.<locals>.decorator_error_code.<locals>.wrapper_error_code(*args, **kwargs)
    571     if error_code > 0:
    572         error_location_msg = (
    573             f'\nError Locations: {error_locations}'
    574             if len(error_locations) > 0
    575             else f''
    576         )
--> 577         raise PrimeRuntimeError(
    578             prime_error_messages.get(
    579                 ErrorCode(error_code), f'Unrecogonized error code {error_code}'
    580             )
    581             + error_location_msg,
    582             ErrorCode(error_code),
    583             error_locations,
    584         )
    586 prime_warnings = []
    587 single_warning = result.get('warningCode', None)

PrimeRuntimeError: CAD import failed.
```

Is there a different way of exporting the fmd files?

**Reply by waltersma** [reply]
Hi @som1197 , 

Can I clarify that once you obtain the FMD file you move it to be visible on the server machine, which I understand to not have a license available for SC, and then attempt the read?  I assume you are also using the absolute file path?  Please verify you are running from a clean virtual environment.  If these assumptions are correct then it is possible there is an issue with your installation.

FMD format is a native CAD format that can be read into pyprimemesh without any out of process readers so you will not need a CAD installation on the server to read this file.  You can export FMD format from SpaceClaim, Discovery or PyAnsys Geometry.  All these options would provide the same result.

Thanks
